File: packages/go/deployment_test.go
package biceptesting

import (
	"context"
	"encoding/json"
	"path/filepath"
	"testing"
	"time"

	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/resources/armdeploymentstacks"
	biceprpcclient "github.com/anthony-c-martin/bicep-testing/packages/go/bicep-rpc-client"
)

func TestDeployCompilesDeploysAndTearsDownOnce(t *testing.T) {
	stackClient := &fakeDeploymentStackClient{}
	originalFactory := newDeploymentStackClient
	newDeploymentStackClient = func(actualSubscriptionID string, credential azcore.TokenCredential) (deploymentStackClient, error) {
		if actualSubscriptionID != subscriptionID {
			t.Errorf("subscription ID = %q, want %q", actualSubscriptionID, subscriptionID)
		}
		return stackClient, nil
	}
	t.Cleanup(func() { newDeploymentStackClient = originalFactory })

	bicep := &fakeBicepClient{compilation: biceprpcclient.CompileParamsResponse{
		Success:    true,
		Template:   `{"resources":[]}`,
		Parameters: `{"parameters":{"message":{"value":"hello"},"secret":{"reference":{"keyVault":{"id":"/subscriptions/sub/resourceGroups/rg/providers/Microsoft.KeyVault/vaults/test"},"secretName":"password","secretVersion":"v1"}}}}`,
	}}
	session := &Session{client: bicep}
	overrides := map[string]json.RawMessage{"message": json.RawMessage(`"override"`)}
	result, err := session.Deploy(context.Background(), fakeCredential{}, DeployOptions{
		FilePath:           "main.bicepparam",
		SubscriptionID:     subscriptionID,
		ResourceGroup:      resourceGroup,
		StackName:          "test-stack",
		ParameterOverrides: overrides,
	})
	if err != nil {
		t.Fatalf("Deploy returned an error: %v", err)
	}
	if !filepath.IsAbs(bicep.request.Path) {
		t.Errorf("compile path = %q, want an absolute path", bicep.request.Path)
	}
	if string(bicep.request.ParameterOverrides["message"]) != `"override"` {
		t.Errorf("parameter override = %s, want override", bicep.request.ParameterOverrides["message"])
	}
	if stackClient.resourceGroup != resourceGroup || stackClient.stackName != "test-stack" {
		t.Errorf("stack target = %s/%s, want %s/test-stack", stackClient.resourceGroup, stackClient.stackName, resourceGroup)
	}
	if actual := stackClient.stack.Properties.Parameters["message"].Value; actual != "hello" {
		t.Errorf("deployment parameter = %v, want hello", actual)
	}
	secret := stackClient.stack.Properties.Parameters["secret"].Reference
	if secret == nil || secret.KeyVault == nil || *secret.KeyVault.ID != "/subscriptions/sub/resourceGroups/rg/providers/Microsoft.KeyVault/vaults/test" || *secret.SecretName != "password" || *secret.SecretVersion != "v1" {
		t.Errorf("secret parameter = %#v, want a Key Vault reference", secret)
	}
	if actual := result.Outputs["endpoint"]; actual != "https://example.test" {
		t.Errorf("endpoint output = %v, want https://example.test", actual)
	}
	if len(result.Resources) != 1 || result.Resources[0].Type != "Microsoft.Storage/storageAccounts" {
		t.Errorf("resources = %#v, want one storage account", result.Resources)
	}

	if err := result.Teardown(context.Background()); err != nil {
		t.Fatalf("Teardown returned an error: %v", err)
	}
	if err := result.Teardown(context.Background()); err != nil {
		t.Fatalf("second Teardown returned an error: %v", err)
	}
	if stackClient.deleteCalls != 1 {
		t.Errorf("delete calls = %d, want 1", stackClient.deleteCalls)
	}
}

type fakeCredential struct{}

func (fakeCredential) GetToken(context.Context, policy.TokenRequestOptions) (azcore.AccessToken, error) {
	return azcore.AccessToken{Token: "token", ExpiresOn: time.Now().Add(time.Hour)}, nil
}

type fakeBicepClient struct {
	request     biceprpcclient.CompileParamsRequest
	compilation biceprpcclient.CompileParamsResponse
}

func (client *fakeBicepClient) CompileParams(_ context.Context, request biceprpcclient.CompileParamsRequest) (biceprpcclient.CompileParamsResponse, error) {
	client.request = request
	return client.compilation, nil
}

func (*fakeBicepClient) GetSnapshot(context.Context, biceprpcclient.GetSnapshotRequest) (biceprpcclient.GetSnapshotResponse, error) {
	return biceprpcclient.GetSnapshotResponse{}, nil
}

func (*fakeBicepClient) Close() error { return nil }

type fakeDeploymentStackClient struct {
	resourceGroup string
	stackName     string
	stack         armdeploymentstacks.DeploymentStack
	deleteCalls   int
}

func (client *fakeDeploymentStackClient) createOrUpdate(_ context.Context, resourceGroup, stackName string, stack armdeploymentstacks.DeploymentStack) (armdeploymentstacks.DeploymentStack, error) {
	client.resourceGroup = resourceGroup
	client.stackName = stackName
	client.stack = stack
	resourceID := "/subscriptions/sub/resourceGroups/rg/providers/Microsoft.Storage/storageAccounts/example"
	return armdeploymentstacks.DeploymentStack{Properties: &armdeploymentstacks.DeploymentStackProperties{
		Outputs: map[string]any{"endpoint": map[string]any{"value": "https://example.test"}},
		Resources: []*armdeploymentstacks.ManagedResourceReference{
			{ID: &resourceID},
		},
	}}, nil
}

func (client *fakeDeploymentStackClient) delete(_ context.Context, resourceGroup, stackName string) error {
	client.deleteCalls++
	return nil
}

File: packages/java/src/test/java/com/github/anthonycmartin/biceptesting/DeploymentTest.java
package com.github.anthonycmartin.biceptesting;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import com.azure.core.credential.AccessToken;
import com.azure.core.credential.TokenCredential;
import com.azure.resourcemanager.resources.deploymentstacks.models.DeploymentStack;
import com.azure.resourcemanager.resources.deploymentstacks.models.DeploymentStackProperties;
import com.azure.resourcemanager.resources.deploymentstacks.models.DeploymentParameter;
import com.azure.resourcemanager.resources.deploymentstacks.models.ManagedResourceReference;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.lang.reflect.Field;
import java.lang.reflect.Proxy;
import java.nio.file.Path;
import java.time.OffsetDateTime;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;
import org.junit.jupiter.api.Test;
import reactor.core.publisher.Mono;

class DeploymentTest {
    private static final ObjectMapper MAPPER = new ObjectMapper();

    @Test
    void deployCompilesDeploysAndTearsDownOnce() throws Exception {
        AtomicInteger deleteCalls = new AtomicInteger();
        DeploymentStack stack = stackResult();
        DeploymentStackService service = new DeploymentStackService() {
            @Override
            public DeploymentStack deploy(
                    String resourceGroup,
                    String stackName,
                    Object template,
                    Map<String, DeploymentParameter> parameters) {
                assertEquals("resource-group", resourceGroup);
                assertEquals("stack", stackName);
                assertEquals("hello", parameters.get("message").value());
                assertEquals("password", parameters.get("secret").reference().secretName());
                assertEquals("v1", parameters.get("secret").reference().secretVersion());
                return stack;
            }

            @Override
            public void delete(String resourceGroup, String stackName) {
                deleteCalls.incrementAndGet();
            }
        };
        var originalFactory = BicepTestSession.deploymentStackServiceFactory;
        BicepTestSession.deploymentStackServiceFactory = (credential, subscription) -> service;
        try {
            FakeRpcCaller rpc = new FakeRpcCaller();
            TokenCredential credential = request -> Mono.just(
                    new AccessToken("token", OffsetDateTime.now().plusHours(1)));
            try (BicepTestSession session = new BicepTestSession(rpc)) {
                DeployResult result = session.deploy(
                        credential,
                        new DeployOptions(
                                Path.of("main.bicepparam"),
                                "subscription",
                                "resource-group",
                                "stack",
                                Map.of("message", MAPPER.valueToTree("override"))));

                assertTrue(Path.of(rpc.params.get("path").asText()).isAbsolute());
                assertEquals("override", rpc.params.get("parameterOverrides").get("message").asText());
                assertEquals("https://example.test", result.outputs().get("endpoint").asText());
                assertEquals("Microsoft.Storage/storageAccounts", result.resources().get(0).type());
                result.close();
                result.close();
            }
        } finally {
            BicepTestSession.deploymentStackServiceFactory = originalFactory;
        }
        assertEquals(1, deleteCalls.get());
    }

    private static DeploymentStack stackResult() throws Exception {
        String resourceId = "/subscriptions/sub/resourceGroups/rg/providers/Microsoft.Storage/storageAccounts/example";
        ManagedResourceReference resource = new ManagedResourceReference();
        setField(resource, "id", resourceId);
        DeploymentStackProperties properties = new DeploymentStackProperties();
        setField(properties, "outputs", Map.of("endpoint", Map.of("value", "https://example.test")));
        setField(properties, "resources", List.of(resource));
        return (DeploymentStack) Proxy.newProxyInstance(
                DeploymentStack.class.getClassLoader(),
                new Class<?>[] {DeploymentStack.class},
                (proxy, method, args) -> method.getName().equals("properties") ? properties : null);
    }

    private static void setField(Object target, String name, Object value) throws Exception {
        Field field = target.getClass().getDeclaredField(name);
        field.setAccessible(true);
        field.set(target, value);
    }

    private static final class FakeRpcCaller implements RpcCaller {
        private JsonNode params;

        @Override
        public JsonNode call(String method, JsonNode params) throws java.io.IOException {
            assertEquals("bicep/compileParams", method);
            this.params = params;
            return MAPPER.readTree("""
                    {
                      "success": true,
                      "template": "{\\"resources\\":[]}",
                                            "parameters": "{\\"parameters\\":{\\"message\\":{\\"value\\":\\"hello\\"},\\"secret\\":{\\"reference\\":{\\"keyVault\\":{\\"id\\":\\"/subscriptions/sub/resourceGroups/rg/providers/Microsoft.KeyVault/vaults/test\\"},\\"secretName\\":\\"password\\",\\"secretVersion\\":\\"v1\\"}}}}"
                    }
                    """);
        }

        @Override
        public void close() {}
    }
}